src/gui/AutomatableModelView.cpp

```diff
@@ -258,7 +279,237 @@ void AutomatableModelViewSlots::removeConnection()
 }
 
 
+void AutomatableModelViewSlots::addSongAutomationNode()
+{
+	// getting all the clips that have this model
+	std::vector<AutomationClip*> clips = AutomationClip::clipsForModel(m_amv->modelUntyped());
+	// selecting the track with the most clips connected to this model
+	AutomationTrack* track = getCurrentAutomationTrack(&clips, true);
+	// getting the clip before the current song time position
+	AutomationClip* clip = getCurrentAutomationClip(track, true, false);
+
+	// getting global song time
+	TimePos timePos = static_cast<TimePos>(Engine::getSong()->getPlayPos());
+	// account for the node's relative position inside clip
+	timePos -= clip->startPosition();
+	bool autoResize = clip->getAutoResize();
+
+	clip->setAutoResize(true);
+	// adding model value
+	clip->recordValue(timePos, m_amv->modelUntyped()->getTrueValue());
+	clip->setAutoResize(autoResize);
+}
+
+void AutomatableModelViewSlots::addSongAutomationNodeAndClip()
+{
+	std::vector<AutomationClip*> clips = AutomationClip::clipsForModel(m_amv->modelUntyped());
+	AutomationTrack* track = getCurrentAutomationTrack(&clips, true);
+	AutomationClip* clip = getCurrentAutomationClip(track, false, false);
+
+	TimePos timePos = static_cast<TimePos>(Engine::getSong()->getPlayPos());
+
+	if (clip && clip->endPosition().getTicks() < timePos.getTicks())
+	{
+		AutomationClip* newClip = makeNewClip(track, timePos, true);
+		// copying the progressionType of the clip before
+		newClip->setProgressionType(clip->progressionType());
+		timePos -= newClip->startPosition();
+		bool autoResize = newClip->getAutoResize();
+
+		newClip->setAutoResize(true);
+		newClip->recordValue(timePos, m_amv->modelUntyped()->getTrueValue());
+		newClip->setAutoResize(autoResize);
+	}
+	else
+	{
+		addSongAutomationNode();
+	}
+}
+
+void AutomatableModelViewSlots::updateSongNearestAutomationNode()
+{
+	std::vector<AutomationClip*> clips = AutomationClip::clipsForModel(m_amv->modelUntyped());
+	// getting the track without adding a new one if no track was found
+	AutomationTrack* track = getCurrentAutomationTrack(&clips, false);
+	// this needs to be checked because getCurrentAutomationTrack might give
+	// a nullptr if it can not find and add a track
+	if (!track) { return; }
+
+	// getting nearest node position
+	AutomationClip* nodeClip = nullptr;
+	TimePos nodePos = getNearestAutomationNode(track, &nodeClip);
+	if (nodeClip)
+	{
+		// modifying its value
+		nodeClip->recordValue(nodePos, m_amv->modelUntyped()->getTrueValue());
+	}
+}
+
+void AutomatableModelViewSlots::removeSongNearestAutomationNode()
+{
+	std::vector<AutomationClip*> clips = AutomationClip::clipsForModel(m_amv->modelUntyped());
+	// getting the track without adding a new one if no track was found
+	AutomationTrack* track = getCurrentAutomationTrack(&clips, false);
```

Comment:
It seems that you could put getting the current `AutomationTrack` into a private helper method as this code repeats through all/most of the new methods. This would make them a bit more readable and remove some "noise".